bundles/org.openhab.binding.tivo/src/main/java/org/openhab/binding/tivo/internal/service/TivoStatusProvider.java

```diff
@@ -0,0 +1,459 @@
+
+        if (connIsConnected()) {
+            logger.debug(" connSocketConnect '{}' - already connected to host '{}', port '{}'",
+                    tivoConfigData.getCfgIdentifier(), tivoConfigData.getHost(), tivoConfigData.getTcpPort());
+            return true;
+        } else {
+            // something is wrong, so force a disconnect/clean up so we can try again
+            connTivoDisconnect();
+        }
+
+        try {
+            Socket tivoSocket = new Socket(tivoConfigData.getHost(), tivoConfigData.getTcpPort());
+            tivoSocket.setKeepAlive(true);
+            tivoSocket.setSoTimeout(CONFIG_SOCKET_TIMEOUT_MS);
+            tivoSocket.setReuseAddress(true);
+
+            if (tivoSocket.isConnected() && !tivoSocket.isClosed()) {
+                if (streamWriter == null) {
+                    streamWriter = new PrintStream(tivoSocket.getOutputStream(), false);
+                }
+                if (this.streamReader == null) {
+                    StreamReader streamReader = new StreamReader(tivoSocket.getInputStream());
+                    streamReader.start();
+                    this.streamReader = streamReader;
+                }
+                this.tivoSocket = tivoSocket;
+            } else {
+                logger.debug(" connSocketConnect '{}' - socket creation failed to host '{}', port '{}'",
+                        tivoConfigData.getCfgIdentifier(), tivoConfigData.getHost(), tivoConfigData.getTcpPort());
+                return false;
+            }
+
+            return true;
+
+        } catch (UnknownHostException e) {
+            logger.debug(" TiVo '{}' - while connecting, unexpected host error: '{}'",
+                    tivoConfigData.getCfgIdentifier(), e.getMessage());
+        } catch (IOException e) {
+            if (tivoStatusData.getConnectionStatus() != ConnectionStatus.OFFLINE) {
+                logger.debug(" TiVo '{}' - I/O exception while connecting: '{}'", tivoConfigData.getCfgIdentifier(),
+                        e.getMessage());
+            }
+        }
+        return false;
+    }
+
+    /**
+     * {@link doNappTime} sleeps for the period specified by the getCmdWaitInterval parameter. Primarily used to allow
+     * the TiVo time to process responses after a command is issued.
+     *
+     * @throws InterruptedException
+     */
+    public void doNappTime() throws InterruptedException {
+        TimeUnit.MILLISECONDS.sleep(tivoConfigData.getCmdWaitInterval());
+    }
+
+    public TivoStatusData getServiceStatus() {
+        return tivoStatusData;
+    }
+
+    public void setServiceStatus(TivoStatusData tivoStatusData) {
+        this.tivoStatusData = tivoStatusData;
+    }
+
+    /**
+     * {@link StreamReader} data stream reader that reads the status data returned from the TiVo.
+     *
+     */
+    public class StreamReader extends Thread {
+        private @Nullable BufferedReader bufferedReader = null;
```

Comment:
Is this thread supposed to be responsible for closing this reader?

Reply:
Yes it should. I added a method to close the reader.